torch/optim/_multi_tensor/_functional.py

```diff
@@ -0,0 +1,267 @@
+        torch._foreach_div_(max_exp_avg_sq_sqrt, bias_correction_sqrt)
+        denom = torch._foreach_add(max_exp_avg_sq_sqrt, eps)
+    else:
+        exp_avg_sq_sqrt = torch._foreach_sqrt(exp_avg_sq)
+        bias_correction_sqrt = [math.sqrt(bc) for bc in bias_correction2]
+        torch._foreach_div_(exp_avg_sq_sqrt, bias_correction_sqrt)
+        denom = torch._foreach_add(exp_avg_sq_sqrt, eps)
+
+    step_size = [(lr / bc) * -1 for bc in bias_correction1]
+    torch._foreach_addcdiv_(params, exp_avg, denom, step_size)
+
+
+def adamax(params: List[Tensor],
+           grads: List[Tensor],
+           exp_avgs: List[Tensor],
+           exp_infs: List[Tensor],
+           states: List[Dict],
+           *,
+           beta1: float,
+           beta2: float,
+           lr: float,
+           weight_decay: float,
+           eps: float):
+    r"""Functional API that performs Adamax algorithm computation.
+
+    See :class:`~torch.optim.Adamax` for details.
+    """
+
+    if weight_decay != 0:
+        torch._foreach_add_(grads, params, alpha=weight_decay)
+
+    # Update biased first moment estimate.
+    torch._foreach_mul_(exp_avgs, beta1)
+    torch._foreach_add_(exp_avgs, grads, alpha=1 - beta1)
+
+    # Update the exponentially weighted infinity norm.
+    torch._foreach_mul_(exp_infs, beta2)
+
+    for exp_inf, grad in zip(exp_infs, grads):
+        norm_buf = torch.cat([
+            exp_inf.unsqueeze(0),
+            grad.abs().add_(eps).unsqueeze_(0)
+        ], 0)
+        torch.max(norm_buf, 0, keepdim=False, out=(exp_inf, exp_inf.new().long()))
+
+    bias_corrections = [1 - beta1 ** state['step'] for state in states]
+    clr = [-1 * (lr / bias_correction) for bias_correction in bias_corrections]
+    torch._foreach_addcdiv_(params, exp_avgs, exp_infs, clr)
+
+
+def adamw(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avg: List[Tensor],
+          exp_avg_sq: List[Tensor],
+          max_exp_avg_sq: List[Tensor],
+          states: List[Dict],
+          *,
+          amsgrad: bool,
+          beta1: float,
+          beta2: float,
+          lr: float,
+          eps: float):
+    r"""Functional API that performs Adamw algorithm computation.
+
+    See :class:`~torch.optim.Adamw` for details.
+    """
+
+    bias_correction1 = [1 - beta1 ** state['step'] for state in states]
+    bias_correction2 = [1 - beta2 ** state['step'] for state in states]
+
+    #
+    # Decay the first and second moment running average coefficient
+    #
+    torch._foreach_mul_(exp_avg, beta1)
+    torch._foreach_add_(exp_avg, grads, alpha=1 - beta1)
+
+    torch._foreach_mul_(exp_avg_sq, beta2)
+    torch._foreach_addcmul_(exp_avg_sq, grads, grads, 1 - beta2)
+
+    if amsgrad:
+        # Maintains the maximum of all 2nd moment running avg. till now
+        for i in range(len(max_exp_avg_sq)):
+            torch.maximum(max_exp_avg_sq[i], exp_avg_sq[i], out=max_exp_avg_sq[i])
```

Comment:
as you pointed out in the original implementation, this was using for_each version of torch maximum. if this only a mypy error that lead to that choice, i recommend looking at `# type: ignore` for now until the issue is resolved upstream.

can you open an issue about the mypy error and link here, and also highlight other places where this may be occurring? would the solution i suggest also be appropriate for the other occurrences?

Reply:
Can you also check at the PRs that created functional multitensor and check if the solution was introduce in them or before?

Reply:
#60766

Reply:
This was a nice offline discussion with Vincent. We concluded to have multi-tensor API's in this PR without mypy typing issue.  However existing optimizers with need to typing will be followed up in another PR.